Userland/Libraries/LibJS/Bytecode/Pass/GenerateCFG.cpp

```diff
@@ -84,60 +87,77 @@ static void generate_cfg_for_block(BasicBlock const& current_block, PassPipeline
             auto finalizer_target = static_cast<Op::EnterUnwindContext const&>(instruction).finalizer_target();
 
             // We keep the frame alive here on the stack, to save some allocation size
-            UnwindFrame frame { nullptr, finalizer_target.has_value() ? &finalizer_target->block() : nullptr };
+            UnwindFrame frame {
+                .handler = nullptr,
+                .finalizer = finalizer_target.has_value() ? &finalizer_target->block() : nullptr
+            };
+
             unwind_frames.append(&frame);
 
+            // We enter the handler target now, to avoid it mutating the unwind stack
+            // without us noticing
+            // FIXME: There probably is a nicer way of doing this
             if (handler_target.has_value()) {
-                // We might pop from this context while in the handler, so we
-                // need to save it and return it to its rightful place
-                // FIXME: We can relax this when we are stricter about entering finally statements
                 auto saved_context = unwind_frames;
+                // FIXME: Not 100% to enter the handler from the current block
                 enter_label(*handler_target, current_block);
-                unwind_frames = move(saved_context);
 
                 frame.handler = &handler_target->block();
-            }
-            {
-                // We might pop from this context while in the handler, so we
-                // need to save it and return it to its rightful place
-                // FIXME: We can relax this when we are stricter about entering finally statements
-                auto saved_context = unwind_frames;
-                enter_label(entry_point, current_block);
                 unwind_frames = move(saved_context);
             }
-            frame.handler = nullptr;
-
-            if (finalizer_target.has_value()) {
-                auto saved_context = unwind_frames;
 
-                enter_label(*finalizer_target, current_block);
-            }
+            // This will enter the handler and finalizer when needed.
+            enter_label(entry_point, current_block);
 
+            if (unwind_frames.last() == &frame)
+                unwind_frames.take_last();
             return;
         }
         case ContinuePendingUnwind: {
             auto resume_target = static_cast<Op::ContinuePendingUnwind const&>(instruction).resume_target();
             enter_label(resume_target, current_block);
+
+            // If we got here following an error, which FIXME: we cannot check statically at the moment
+            // we need to enter the next unwind frames hander, which is marked in the general,
+            // but if that one does not exist we need to enter its finalizer
```

Comment:
This comment is awkward to read and has typos.
"which is marked in the general" what?
Also please put the FIXME into it's own line, it makes this statement harder to read when it is in the middle of a sentence. Something like:
```
// If we got here following an error, we need to enter the next unwind frames handler, which is
// marked in the general *what?*. If that handler does not exist, we need to enter its finalizer.
// FIXME: We cannot statically check *what exactly?* at the moment.
```